enhancements/baremetal/baremetal-poison-pill-remediation.md

```diff
@@ -0,0 +1,189 @@
+If there’s a network outage between the node running CAPBM (which expects to use BMC commands like power-off etc.) and the
+unhealthy node, the power-cycle attempt cannot
+succeed. Self health checks and remediation can resolve such cases, however that work/scenario is out of scope for
+this proposal.
+
+## Motivation
+
+Some clusters don’t have BMC credentials and we still want to have auto remediation for failing nodes.
+This proposal allows:
+1. To allow stateful workload running on the unhealthy nodes to be rescheduled after finite time on other healthy nodes.
+1. To restore compute capacity by remediating the unhealthy node
+
+### Goals
+
+* To allow remediation of bare metal nodes without BMC credentials
+* To utilize the MHC logic for deciding if a node is unhealthy and whether remediation is appropriate
+* To allow the healthy majority to recover workloads after a pre-determined and finite interval
+* To allow a node to take actions to remediate itself in case of a failure
+* To allow a node to detect its health status when master<->node communication is partially failed
+* To avoid false positives caused by failure of the control plane
+
+### Non-Goals
+
+* Self healing in combination with existing reprovisioning or reboot based remediation mechanisms in the Machine Health Check controller
+* Recovering from all types of failure
+* Having a distributed mesh of health checks where each node is checking health for its peers
+* Creation of fake Machine objects for UPI clusters
+
+## Proposal
+
+Utilizing a DaemonSet to have a privileged critical pod running on each node. This pod will periodically check the
+health status of the node it’s running on, and tickle a [watchdog device](https://en.wikipedia.org/wiki/Watchdog_timer).
+
+This will be performed by looking for the `host.metal3.io/external-remediation` annotation added by MHC to the unhealthy
+Machine CR.
+
+If the unhealthy annotation was found, the node will remediate itself by using the watchdog device to
+trigger a reset of the node. Watchdog devices, especially the hardware kind, are considered more reliable and predictable than the
+reboot command.
+
+If Machine CR is not accessible in 3 consecutive attempts, it needs external assistance to detect its
+health status.
+
+Failure to read Machine CR could stem from one of the following:
+1. The node is unhealthy and therefore can’t read the Machine CR (e.g. local network failure)
+1. There’s an etcd failure
+1. Connectivity issues between the node and the control plane nodes (which doesn’t stem from local problem in the node)
+1. Resource starvation, preventing our DaemonSet from initiating the check 
+1. Authentication failure to API (ServiceAccount deleted?)
+1. RBAC disallows read operation on Machine (Role deleted or changed
+
+We only want to remediate if the problem is local (#1 and #4). Otherwise, we could create a "remediation storm" where
+all nodes are trying to remediate themselves, even if the problem is not theirs.
```

Comment:
#1 from #2 - using other peers. if other peers can access etcd, we know that it's not #2. if other peers can't access etcd we assume it's #2. if the failing node can't reach his peers, nor access etcd we assume it's #1.

#1 from #3 - using other peers to differentiate. if other peers can reach etcd we assume it's #3, if the suspected node can't reach etcd, nor any of its peers we assume it's #1.

Reply:
> If we can detect #1, #3 and #4 why aren't the nodes throwing alerts that can be used for other debugging purposes?

if it's #1, the node probably won't be able to send alerts anywhere (e.g. if it has local network failure)
To identify #3 we had to introduce communication between the peers. We could identify this without them. 
as for #4, there are some existing node conditions such as mem/disk pressure that are already reported. We use watchdog device since we must make sure that the poison pill pod is running or the machine is not running any workloads. Otherwise we're risking in running staeful sets with run-once semantics in two different nodes.